eclair-core/src/main/scala/fr/acinq/eclair/payment/MultiPartPaymentLifecycle.scala

```diff
@@ -0,0 +1,407 @@
+      // We tell the router to compute those stats though: in case our payment attempt fails, they will be available for
+      // another payment attempt.
+      if (s.stats.isEmpty) {
+        router ! TickComputeNetworkStats
+      }
+      relayer ! GetOutgoingChannels()
+      goto(PAYMENT_IN_PROGRESS) using PaymentProgress(d.sender.get, r, s.stats, r.totalAmount, r.maxAttempts, Map.empty, Nil)
+  }
+
+  when(PAYMENT_IN_PROGRESS) {
+    case Event(OutgoingChannels(channels), d: PaymentProgress) if d.toSend > 0.msat =>
+      log.debug("trying to send {} with local channels: {}", d.toSend, channels.map(_.toUsableBalance).mkString(","))
+      val randomize = d.remainingAttempts != d.request.maxAttempts // we randomize channel selection when we retry
+      val (remaining, payments) = splitPayment(nodeParams, d.toSend, channels, d.networkStats, d.request, randomize)
+      if (remaining > 0.msat) {
+        log.warning(s"cannot send ${d.toSend} with our current balance")
+        goto(PAYMENT_ABORTED) using PaymentAborted(d.sender, d.request, d.failures :+ LocalFailure(new RuntimeException("balance is too low")), d.pending.keySet)
+      } else {
+        val pending = setFees(d.request.routeParams, payments, payments.size + d.pending.size)
+        pending.foreach { case (childId, payment) => spawnChildPaymentFsm(childId) ! payment }
+        stay using d.copy(toSend = 0 msat, remainingAttempts = d.remainingAttempts - 1, pending = d.pending ++ pending)
+      }
+
+    case Event(pf: PaymentFailed, d: PaymentProgress) =>
+      val paymentTimedOut = pf.failures.exists {
+        case f: RemoteFailure => f.e.failureMessage == PaymentTimeout
+        case _ => false
+      }
+      if (paymentTimedOut) {
+        goto(PAYMENT_ABORTED) using PaymentAborted(d.sender, d.request, d.failures ++ pf.failures, d.pending.keySet - pf.id)
+      } else if (d.remainingAttempts == 0) {
+        val failure = LocalFailure(new RuntimeException("payment attempts exhausted without success"))
+        goto(PAYMENT_ABORTED) using PaymentAborted(d.sender, d.request, d.failures ++ pf.failures :+ failure, d.pending.keySet - pf.id)
+      } else {
+        // Get updated local channels (will take into account the child payments that are in-flight).
+        if (d.toSend == 0.msat) relayer ! GetOutgoingChannels()
+        val failedPayment = d.pending(pf.id)
+        stay using d.copy(toSend = d.toSend + failedPayment.finalPayload.amount, pending = d.pending - pf.id, failures = d.failures ++ pf.failures)
+      }
+
+    case Event(ps: PaymentSent, d: PaymentProgress) =>
+      require(ps.parts.length == 1, "child payment must contain only one part")
+      // As soon as we get the preimage we can consider that the whole payment succeeded (we have a proof of payment).
+      goto(PAYMENT_SUCCEEDED) using PaymentSucceeded(d.sender, d.request, ps.paymentPreimage, ps.parts, d.pending.keySet - ps.id)
+  }
+
+  when(PAYMENT_ABORTED) {
```

Comment:
I'm not sure the `PAYMENT_ABORTED`/`PAYMENT_SUCCEEDED` distinction really make sense, because, as you have noticed, we can receive a payment proof even in the failure state.

Reply:
But if we have a single state for both we'll have ugly if/switch everywhere to be able to handle the "spec violation" case (and it feels a bit cleaner to express the difference via different states, because you can move from `PAYMENT_ABORTED` to `PAYMENT_SUCCEEDED` but not the other way around.